testcases/cve/stack_clash.c

```diff
@@ -0,0 +1,251 @@
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV)
+		return;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO,
+		"mmap = [%lx, %lx), addr = %lx, diff = %lx, THRESHOLD = %lx",
+		mapped_addr, mmap_end, fault_addr, diff, THRESHOLD);
+	if (diff < 0 || (unsigned long)diff < THRESHOLD)
+		_exit(EXIT_FAILURE);
+	else
+		_exit(EXIT_SUCCESS);
+}
+
+unsigned long read_stack_addr_from_proc(void)
+{
+	FILE *fh;
+	char buf[1024];
+	unsigned long stack_top = -1UL, start, end;
+
+	fh = SAFE_FOPEN("/proc/self/maps", "r");
+
+	while (!feof(fh)) {
+		if (fgets(buf, sizeof(buf), fh) == NULL) {
+			tst_brk(TBROK | TERRNO, "fgets");
+			goto out;
+		}
+
+		if (!strstr(buf, "[stack"))
+			continue;
+
+		if (sscanf(buf, "%lx-%lx", &start, &end) != 2) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			goto out;
+		}
+
+		if (STACK_GROWSDOWN)
+			stack_top = start;
+		else
+			stack_top = end;
+		break;
+	}
+
+out:
+	SAFE_FCLOSE(fh);
+	return stack_top;
+}
+
+void do_child(void)
+{
+	unsigned long stack_addr;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	void *map;
+	unsigned long gap = GAP_PAGES * PAGE_SIZE;
+
+	stack_addr = read_stack_addr_from_proc();
+	if (stack_addr == -1UL) {
+		tst_brk(TBROK, "can't read stack top from /proc/self/maps");
+		return;
+	}
+
+	if (STACK_GROWSDOWN)
+		mapped_addr = stack_addr - gap - MAPPED_LEN;
+	else
+		mapped_addr = stack_addr + gap;
+
+	mapped_addr &= PAGE_MASK;
+	map = SAFE_MMAP((void *)mapped_addr, MAPPED_LEN,
+			PROT_READ|PROT_WRITE,
+			MAP_ANON|MAP_PRIVATE|MAP_FIXED, -1, 0);
+	tst_res(TINFO, "Stack top:0x%lx mmap:%p+0x%lx", stack_addr, map,
+		MAPPED_LEN);
+
+	signal_stack.ss_sp = SAFE_MALLOC(SIGNAL_STACK_SIZE);
+	signal_stack.ss_size = SIGNAL_STACK_SIZE;
+	signal_stack.ss_flags = 0;
+	if (sigaltstack(&signal_stack, NULL) == -1) {
+		tst_brk(TBROK | TERRNO, "sigaltstack");
+		return;
+	}
+	if (sigaction(SIGSEGV, &segv_sig, NULL) == -1 ||
+	    sigaction(SIGBUS,  &segv_sig, NULL) == -1) {
+		tst_brk(TBROK | TERRNO, "sigaction");
+		return;
+	}
+
+#ifdef DEBUG
+	{
+		char buf[128];
+		const char *cmd[] = {"cat", buf, NULL};
+		sprintf(buf, "/proc/%d/maps", getpid());
+		tst_run_cmd(cmd, NULL, NULL, 0);
+	}
+#endif
+
+	exhaust_stack_into_sigsegv();
+}
+
+void setup(void)
+{
+	char buf[1024], *p;
+	FILE *fh;
+
+	PAGE_SIZE = sysconf(_SC_PAGESIZE);
+	PAGE_MASK = ~(PAGE_SIZE - 1);
+
+
+	fh = SAFE_FOPEN("/proc/cmdline", "r");
+	if (fgets(buf, sizeof(buf), fh) == NULL) {
+		tst_brk(TBROK | TERRNO, "fgets");
+		return;
+	}
+
+
+	if ((p = strstr(buf, "stack_guard_gap=")) != NULL) {
+		if (sscanf(p, "stack_guard_gap=%ld", &GAP_PAGES) != 1) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			return;
+		}
+		tst_res(TINFO, "stack_guard_gap = %ld", GAP_PAGES);
+	}
```

Comment:
I think `SAFE_FILE_SCANF` in `safe_file_ops.h` should be able to read `/proc/cmdline` in one line.